File: 0874-walking-robot-simulation/0874-walking-robot-simulation.ts
function robotSim(commands: number[], obstacles: number[][]): number {
    let st =  new Set<string>();

    for(let obs of obstacles){
        let key = obs[0].toString() + "_" + obs[1].toString();
        st.add(key);
    }

    let x=0;
    let y=0;
    let ans=0;

    let [dirx,diry] = [0,1];

    for(let i=0;i<commands.length;i++){
        if(commands[i]===-2){
            [dirx,diry] = [-diry,dirx]; //turning left
        }
        else if(commands[i]===-1){
            [dirx,diry] = [diry,-dirx]; //turning right
        }
        else{
            for(let j=0;j<commands[i];j++){
                let newX = x+dirx;
                let newY = y+diry;

                let nextKey = newX.toString() +"_"+ newY.toString();
                if(st.has(nextKey)){
                    break;
                }
                x=newX;
                y=newY;
            }
            
        }
        ans=Math.max(ans,x*x+y*y);
    }

    return ans;
};

File: 0874-walking-robot-simulation/0874-walking-robot-simulation.cpp
class Solution {
public:
    int robotSim(vector<int>& commands, vector<vector<int>>& obstacles) {
        unordered_set<string> st;

        for(int i=0;i<obstacles.size();i++){
            string key = to_string(obstacles[i][0]) + "_" + to_string(obstacles[i][1]);
            st.insert(key);
        }

        int x=0;
        int y=0;
        int maxD=0;

        pair<int,int> dir ={0,1};

        for(int i=0;i<commands.size();i++){
            if(commands[i]==-2){
                dir={-dir.second,dir.first};
            }
            else if(commands[i]==-1){
                dir={dir.second,-dir.first};
            }
            else{
                for(int j=0;j<commands[i];j++){
                    int newX=x+dir.first;
                    int newY=y+dir.second;
                    string nextKey = to_string(newX)+"_"+ to_string(newY);
                    if(st.find(nextKey)!= st.end()){
                        break;
                    }
                    x=newX;
                    y=newY;
                }
            }
            maxD=max(maxD,x*x+y*y);
        }

        return maxD;
    }
};